src/frontend/pages/article/table_of_contents.rs

```diff
@@ -0,0 +1,160 @@
+                    return;
+                }
+            }
+        }
+    }
+
+    pub(crate) fn push(&mut self, level: u32, name: String, id: String) -> &str {
+        assert!(level >= 1);
+
+        self.fold_until(level);
+
+        let mut sec_number;
+        {
+            let toc = match self.chain.last() {
+                None => {
+                    sec_number = String::new();
+                    &self.top_level
+                }
+                Some(entry) => {
+                    sec_number = entry.sec_number.clone();
+                    sec_number.push('.');
+                    &entry.children
+                }
+            };
+            let number = toc.count_entries_with_level(level);
+            sec_number.push_str(&(number + 1).to_string())
+        }
+
+        self.chain.push(TocEntry {
+            level,
+            name,
+            sec_number,
+            id,
+            children: Toc {
+                entries: Vec::new(),
+            },
+        });
+
+        let just_inserted = self.chain.last_mut().unwrap();
+        &just_inserted.sec_number
+    }
+}
+
+impl Toc {
+    fn print_inner(&self, v: &mut String) {
+        use std::fmt::Write as _;
+
+        if !&self.entries.is_empty() {
+            v.push_str("<ul>");
+            for entry in &self.entries {
+                let _ = write!(
+                    v,
+                    "\n<li><a href=\"#{id}\">{num} {name}</a>",
+                    id = entry.id,
+                    num = entry.sec_number,
+                    name = entry.name
+                );
+                entry.children.print_inner(&mut *v);
+                v.push_str("</li>");
+            }
+            v.push_str("</ul>");
+        }
+    }
+    fn print(&self) -> String {
+        let mut v = String::from("<li class=\"menu-title\">Table of Contents</li>");
+        self.print_inner(&mut v);
+        v
+    }
+}
+
+pub fn generate_table_of_contents(text: &str) -> String {
+    let mut toc_builder = TocBuilder::new();
+    text.lines()
+        .filter(|x| {
+            if !x.starts_with("#") {
+                return false;
+            }
+            x.chars()
+                .skip_while(|char| *char == '#')
+                .collect::<String>()
+                .starts_with(" ")
+        })
+        .for_each(|x| {
+            let mut level: u32 = 0;
+            let line = x
+                .chars()
+                .skip_while(|x| {
+                    level += 1;
+                    *x == '#' || *x == ' '
+                })
+                .collect::<String>();
+            toc_builder.push(level - 2, line.clone().to_string(), to_kebab_case(line));
+        });
+    let toc = toc_builder.into_toc();
+    toc.print()
+}
+
+pub fn to_kebab_case(line: String) -> String {
+    line.to_lowercase()
+        .chars()
+        .filter(|x| x.is_alphabetic() || *x == ' ')
+        .collect::<String>()
+        .replace(" ", "-")
+}
```

Comment:
This logic is wrong, for example headings which have only numbers end up with an empty id. Checkout the code for the [heading anchors plugin](https://github.com/markdown-it-rust/markdown-it-plugins.rs/blob/main/crates/heading_anchors/src/lib.rs#L133). It uses [github-slugger](https://crates.io/crates/github-slugger) so you should do the same for consistency.